### PR TITLE
Add ability to find config files in multiple dirs

### PR DESCRIPTION
**Pull request by veerajsk**
Wrike Task: https://www.wrike.com/open.htm?id=1061561076

- The engine now has a set of directories where it looks for config files recursively
- If the roleset with the given file name is not found (recursive search), then the engine looks for any roleset file in the specified directories but not recursively
- The beh & cnd files are now referred to only by their file name & not by their relative path to the plans folder
- The plan, role & task config keys are no longer required since the search is recursive & hence these are removed

### CODE REVIEW

File: alica_engine/include/engine/AlicaContext.h
Comment (ntfshard) [inline alica_engine/include/engine/AlicaContext.h]: maybe we want to have user-defined order instead of unpredictable unordered_set's order

File: alica_engine/src/engine/AlicaContext.cpp
Comment (ntfshard) [inline alica_engine/src/engine/AlicaContext.cpp]: Return node even if error occurred? Perhaps we should continue search?

File: alica_engine/src/engine/AlicaContext.cpp
Comment (ntfshard) [inline alica_engine/src/engine/AlicaContext.cpp]: copy-paste

File: supplementary/alica_ros2/alica_ros2_turtlesim/src/base.cpp
Comment (ntfshard) [inline supplementary/alica_ros2/alica_ros2_turtlesim/src/base.cpp:34]: consider use smart pointer (several places)

File: alica_engine/include/engine/modelmanagement/ModelManager.h
Comment (ntfshard) [inline alica_engine/include/engine/modelmanagement/ModelManager.h]: const?

File: alica_engine/include/engine/modelmanagement/ModelManager.h
Comment (ntfshard) [inline alica_engine/include/engine/modelmanagement/ModelManager.h]: const?

File: alica_engine/src/engine/modelmanagement/ModelManager.cpp
Comment (ntfshard) [inline alica_engine/src/engine/modelmanagement/ModelManager.cpp:57]: `push_back(fileToParse)`?

File: alica_engine/src/engine/modelmanagement/ModelManager.cpp
Comment (ntfshard) [inline alica_engine/src/engine/modelmanagement/ModelManager.cpp]: by copy?

**Review by ntfshard — COMMENTED** [review]
New code not covered by tests, correct me if I wrong

**Comment by ntfshard** [comment]
and in general, I guess we should start using https://en.cppreference.com/w/cpp/filesystem

File: alica_engine/include/engine/AlicaContext.h
Comment (veerajsk) [reply alica_engine/include/engine/AlicaContext.h]: yeah, that would be better, I'll change to std::vector

File: alica_engine/src/engine/AlicaContext.cpp
Comment (veerajsk) [reply alica_engine/src/engine/AlicaContext.cpp]: yes, I'll remove the return

File: supplementary/alica_ros2/alica_ros2_turtlesim/src/base.cpp
Comment (veerajsk) [reply supplementary/alica_ros2/alica_ros2_turtlesim/src/base.cpp:34]: would prefer making unrelated changes in a separate PR & address all places where smart pointers can be used in that PR

File: alica_engine/include/engine/modelmanagement/ModelManager.h
Comment (veerajsk) [reply alica_engine/include/engine/modelmanagement/ModelManager.h]: I think it is better to return non const if returning by value for 2 reasons:
1. Returning by const can prevent moving from the returned object which unnecessarily hurts performance eg. ` file = findFile` will always assigns by copy (instead of move) if returned by const
2. The returned value is anyway in the caller's scope & I prefer the caller to have the flexibility to own a non-const version & convert to const if he does really need it.

File: alica_engine/src/engine/modelmanagement/ModelManager.cpp
Comment (veerajsk) [reply alica_engine/src/engine/modelmanagement/ModelManager.cpp:57]: fileToParse is the complete path where filesToParse.front() is just the name of the file

File: alica_engine/src/engine/modelmanagement/ModelManager.cpp
Comment (veerajsk) [reply alica_engine/src/engine/modelmanagement/ModelManager.cpp]: I'll change it to reference

**Comment by veerajsk** [comment]
> and in general, I guess we should start using https://en.cppreference.com/w/cpp/filesystem

I agree. But again, I would defer to another PR in the interest of keeping this PR small & focused

**Comment by veerajsk** [comment]
> New code not covered by tests, correct me if I wrong

There is a test for parsing the plan files already & it verifies this change as well for a single folder config path. I am adding another test for multiple folders (PR is still marked draft for this reason)

File: alica_engine/include/engine/modelmanagement/ModelManager.h
Comment (ntfshard) [reply alica_engine/include/engine/modelmanagement/ModelManager.h]: I meant const qualifier of member function

File: alica_engine/include/engine/modelmanagement/ModelManager.h
Comment (ntfshard) [reply alica_engine/include/engine/modelmanagement/ModelManager.h]: const qualifier of member function

File: alica_engine/src/engine/AlicaContext.cpp
Comment (ntfshard) [inline alica_engine/src/engine/AlicaContext.cpp]: This is a same error messages, perhapce using variable could help to avoid copy-paste

File: alica_engine/src/engine/AlicaContext.cpp
Comment (ntfshard) [inline alica_engine/src/engine/AlicaContext.cpp]: redundant variable

File: alica_engine/src/engine/AlicaContext.cpp
Comment (ntfshard) [reply alica_engine/src/engine/AlicaContext.cpp]: Same below

File: alica_engine/src/engine/modelmanagement/ModelManager.cpp
Comment (ntfshard) [inline alica_engine/src/engine/modelmanagement/ModelManager.cpp]: variable scope can be reduced; technically you can use `auto node = YAML::LoadFile(file);`

File: alica_engine/src/engine/modelmanagement/factories/Factory.cpp
Comment (ntfshard) [inline alica_engine/src/engine/modelmanagement/factories/Factory.cpp:64]: perhaps else branch also should be here

File: alica_engine/src/engine/AlicaContext.cpp
Comment (Dale-Koenig) [reply alica_engine/src/engine/AlicaContext.cpp]: IMO maybe logging component itself should output a generic msg if it is used pre-initialization rather than having custom error msgs in multiple places

Comment (veerajsk) [reply alica_engine/src/engine/AlicaContext.cpp]: yes, I can take a look into implementing something like this in the logging component, because we have this issue in other places as well eg. some user code using the logger before it is initialised. I have observed this when other engine components are set before the logger is set

**Review by bjoernschroeder — COMMENTED** [review]
I dont have much to add to the already made review comments. PR is LGTM after resolving them.

**Review by Dale-Koenig — COMMENTED** [review]
Please make sure lbc+lbc bridge work without any changes and add backwards compatible (deprecated) functions where necessary

File: alica_engine/src/engine/modelmanagement/factories/Factory.cpp
Comment (veerajsk) [reply alica_engine/src/engine/modelmanagement/factories/Factory.cpp:64]: else is not required, because there is nothing to do in it. Basically the else block will be hit when the file has a valid extension but is already either parsed or included in the list of files to be parsed. Therefore nothing needs to be done in either case.

**Review by Dale-Koenig — APPROVED** [review]

File: alica_engine/src/engine/modelmanagement/ModelManager.cpp
Comment (ntfshard) [inline alica_engine/src/engine/modelmanagement/ModelManager.cpp]: technically you can initialize variable on creation
```suggestion
        const auto fileToParse = findFile(filesToParse.front());
        if (fileToParse.empty()) {
```

File: alica_engine/src/engine/modelmanagement/ModelManager.cpp
Comment (ntfshard) [inline alica_engine/src/engine/modelmanagement/ModelManager.cpp]:
```suggestion
    auto roleSetPath = findFile(roleSetName + alica::Strings::roleset_extension);
```
and you don't need such long expression inside `if`

File: supplementary/alica_ros1/alica_tests/src/test/test_alica_engine_plan_parser.cpp
Comment (ntfshard) [inline supplementary/alica_ros1/alica_tests/src/test/test_alica_engine_plan_parser.cpp:259]: can't review this changes properly

**Review by ntfshard — COMMENTED** [review]
I guess it's ok

File: supplementary/alica_ros1/alica_tests/src/test/test_alica_engine_plan_parser.cpp
Comment (veerajsk) [reply supplementary/alica_ros1/alica_tests/src/test/test_alica_engine_plan_parser.cpp:259]: I didn't make any changes to this code. The only thing I did is to move the plan parsing checks into a function in the base class & call it in both tests

**Comment by veerajsk** [comment]
> I guess it's ok

I addressed all the review comments, so could you please approve if it's ok?

**Review by ntfshard — APPROVED** [review]

File: alica_engine/src/engine/modelmanagement/ModelManager.cpp
Comment (ntfshard) [inline alica_engine/src/engine/modelmanagement/ModelManager.cpp:43]: same case here, if you want

Comment (veerajsk) [reply alica_engine/src/engine/modelmanagement/ModelManager.cpp:43]: I think this is nit & we need this change to unblock changes in the backend. So merging

**Pull request merged**
